gapis/api/sync/lookup.go

```diff
@@ -0,0 +1,249 @@
+		l.commandBuffers[commandBuffer] = cbl
+	}
+	cbl.add(submission, idx)
+}
+
+// AddRenderPass adds a submitted render pass start index to the mapping.
+func (l *RenderPassLookup) AddRenderPass(ctx context.Context, key RenderPassKey, idx SubCmdRange) {
+	log.D(ctx, "Adding mapping for render pass %v -> [%v, %v]", key, idx.From, idx.To)
+	rpl, ok := l.renderPasses[key.RenderPass]
+	if !ok {
+		rpl = newRenderPassLookup()
+		l.renderPasses[key.RenderPass] = rpl
+	}
+	rpl.add(key, idx)
+}
+
+// Lookup finds the best matching command index for the given key. Specifying zero for any of the
+// handles is treated as "unknown" and will cause the lookup to match up with the best known
+// index, if it exists. Returned indecies either point to a submitted command buffer or a render
+// pass within a submitted command buffer.
+func (l *RenderPassLookup) Lookup(ctx context.Context, key RenderPassKey) SubCmdRange {
+	if key.RenderPass != 0 {
+		if rpl, ok := l.renderPasses[key.RenderPass]; ok {
+			return rpl.lookup(key)
+		}
+	}
+	if cbl, ok := l.commandBuffers[key.CommandBuffer]; ok {
+		return cbl.lookup(key)
+	}
+	return SubCmdRange{}
+}
+
+type commandBufferLookup struct {
+	submissions     map[int]api.SubCmdIdx
+	firstSubmission int
+}
+
+func newCommandBufferLookup() *commandBufferLookup {
+	return &commandBufferLookup{
+		submissions:     map[int]api.SubCmdIdx{},
+		firstSubmission: math.MaxInt,
+	}
+}
+
+func (l *commandBufferLookup) add(submission int, idx api.SubCmdIdx) {
+	if current, ok := l.submissions[submission]; !ok || idx.LessThan(current) {
+		l.submissions[submission] = idx
+	}
+	if submission < l.firstSubmission {
+		l.firstSubmission = submission
+	}
+}
+
+func (l *commandBufferLookup) lookup(key RenderPassKey) SubCmdRange {
+	if idx, ok := l.submissions[key.Submission]; ok {
+		return SubCmdRange{idx, idx}
+	}
+
+	if key.Submission == 0 || len(l.submissions) == 1 {
+		idx := l.submissions[l.firstSubmission]
+		return SubCmdRange{idx, idx}
+	}
+
+	// Find the command buffer that matches the submission index the closest.
+	var idx api.SubCmdIdx
+	distance := math.MaxInt
+	for submission, cmd := range l.submissions {
+		d := abs(key.Submission - submission)
+		if d < distance {
+			idx = cmd
+			distance = d
+		}
+	}
+	return SubCmdRange{idx, idx}
+}
+
+type renderPassLookup struct {
+	mappings        map[RenderPassKey]SubCmdRange
+	byCommandBuffer map[uint64][]RenderPassKey
+	byFramebuffer   map[uint64][]RenderPassKey
+}
+
+func newRenderPassLookup() *renderPassLookup {
+	return &renderPassLookup{
+		mappings:        map[RenderPassKey]SubCmdRange{},
+		byCommandBuffer: map[uint64][]RenderPassKey{},
+		byFramebuffer:   map[uint64][]RenderPassKey{},
+	}
+}
+
+func (l *renderPassLookup) add(key RenderPassKey, idx SubCmdRange) {
+	key.RenderPass = 0 // clear it out, since we don't care about it anymore.
+
+	if current, ok := l.mappings[key]; !ok || !current.contains(idx) {
+		l.mappings[key] = current.expand(idx)
```

Comment:
My refactor suggestion is based on that if a key cannot be found in a map, it would return `nil` and if that assumption is wrong, therefore it is expected that the refactor will not do the right thing. It does not have anything about the code or the code style. 

I would not expect a value returned by a non-existing element but if that's what Go does the let's make sure that the returned element, whatever it is, suits to the logic.

And for your comment, I can't see how it prevents easily adding something at the end of the function.

Reply:
Even with that assumption, the re-writes are broken as they are not updating `l.mappings` in the `!ok` case, making the lookup redundant as there never will be any mappings. My point is that early return doesn't inherently make code more readable or avoid control-flow bugs any more than non-early return style.

re: adding stuff to the end of a function. If I have a block that may contain a bunch of `if`s to skip parts of it, control flow will still continue after the block. While, if I go ahead and change it to do early returns instead of skips, control flow will not always reach the end of the block. If the early return is a bit buried, someone might miss it and wonder why new code added after the block is not run in some circumstances. Don't get me wrong, this is a minor issue in most cases and I'm fine with early returns.